### PR TITLE
Dispatch error event when link elements get invalid response

### LINKED ISSUE #10374: Dispatch error event when link elements get invalid network response
Specification: https://html.spec.whatwg.org/multipage/semantics.html#concept-link-obtain (starting at "Once the attempts to obtain the resource")
Code: `components/script/dom/htmllinkelement.rs` (see `AsyncResponseListener` implementation for `StylesheetContext`)
Test: `./mach test-wpt tests/wpt/web-platform-tests/html/semantics/document-metadata/the-link-element/link-style-error-01.html`

We should check whether the status received in `response_complete` is an error, and also check whether the Metadata value stored in `headers_available` contains a `Content-Type` value which is anything besides `text/stylesheet`. If either of these are true, we should call `fire_simple_event` on the original `HTMLLinkElement` element.

cc @nehagup

### PR DESCRIPTION
**Pull request by askeing**
fix #10374

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10887)

### CODE REVIEW

**Comment by askeing** [comment]
From the beginning, I try to match the `TopLevel::Text, SubLevel::Ext("stylesheet")` for `text/stylesheet`.
But I found [RFC2318](https://tools.ietf.org/html/rfc2318), it seems like we should check the `text/css`, not `text/stylesheet`.
So I will change to match `TopLevel::Text, SubLevel::Css`.

File: tests/wpt/metadata/html/semantics/document-metadata/the-link-element/link-style-error-01.html.ini [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [link-style-error-01.html]
   type: testharness
-  expected: TIMEOUT
+  expected: OK
   [Should get an error event for a text/plain response.]
-    expected: TIMEOUT
+    expected: PASS
```
Comment (KiChjang) [inline tests/wpt/metadata/html/semantics/document-metadata/the-link-element/link-style-error-01.html.ini]: By default, test files are expected to PASS, so you can instead remove this file.

File: components/script/dom/htmllinkelement.rs [hunk 2]
```diff
@@ -281,6 +291,13 @@ impl AsyncResponseListener for StylesheetContext {
     }
 
     fn response_complete(&mut self, _status: Result<(), NetworkError>) {
+        match _status {
```
Comment (KiChjang) [inline components/script/dom/htmllinkelement.rs]: `if status.is_err() {`

File: components/script/dom/htmllinkelement.rs [hunk 3]
```diff
@@ -281,6 +291,13 @@ impl AsyncResponseListener for StylesheetContext {
     }
 
     fn response_complete(&mut self, _status: Result<(), NetworkError>) {
```
Comment (KiChjang) [inline components/script/dom/htmllinkelement.rs]: You can rename `_status` to `status`.

File: components/script/dom/htmllinkelement.rs [hunk 4]
```diff
@@ -273,6 +276,13 @@ impl PreInvoke for StylesheetContext {}
 impl AsyncResponseListener for StylesheetContext {
     fn headers_available(&mut self, metadata: Result<Metadata, NetworkError>) {
         self.metadata = metadata.ok();
+        let content_type = self.metadata.clone().and_then(|meta| meta.content_type);
```
Comment (KiChjang) [inline components/script/dom/htmllinkelement.rs]: Is this clone necessary?

File: components/script/dom/htmllinkelement.rs [hunk 5]
```diff
@@ -273,6 +276,13 @@ impl PreInvoke for StylesheetContext {}
...
+        match content_type {
+            Some(ContentType(Mime(TopLevel::Text, SubLevel::Css, _))) => {},
```
Comment (KiChjang) [inline components/script/dom/htmllinkelement.rs]:
```rust
if let Some(ContentType(Mime(TopLevel::Text, SubLevel::Css, _))) = content_type {
} else {
    self.elem.root.upcast::<EventTarget>().fire_simple_event("error");
}
```

File: components/script/dom/htmllinkelement.rs (hunk 4 above)
Comment (askeing) [inline components/script/dom/htmllinkelement.rs]: you are right, it seems like I can use the `if let` and `ref` to get the `content_type`.

File: components/script/dom/htmllinkelement.rs [hunk 6]
```diff
@@ -273,14 +276,24 @@ impl PreInvoke for StylesheetContext {}
...
     }
 
     fn data_available(&mut self, payload: Vec<u8>) {
         let mut payload = payload;
         self.data.append(&mut payload);
     }
 
-    fn response_complete(&mut self, _status: Result<(), NetworkError>) {
+    fn response_complete(&mut self, status: Result<(), NetworkError>) {
+        if status.is_err() {
```
Comment (Ms2ger) [inline components/script/dom/htmllinkelement.rs]: I don't think this case can happen right now, so I'd rather you made this `assert!(status.is_ok());` for now.

Comment (askeing) [inline components/script/dom/htmllinkelement.rs]: Why the NetworkError doesn't happen here right now?
After adding the `assert!(status.is_ok());` here, the assertion will failed when I try to disable the network connection, and open a local HTML file with remote CSS file.

**Comment by KiChjang** [comment]
@bors-servo r+

Thanks!

**Comment by Ms2ger** [comment]
https://github.com/servo/servo/issues/10921

**Pull request merged**